[flang][runtime] Build ISO_FORTRAN_ENV to export kind arrays as linkable symbols

mjklemm opened the pull request:
Moves definitions of the kind arrays into a Fortran MODULE to not only emit the MOD file, but also compile that MODULE file into an object file.  This file is then linked into libFortranRuntime.so.

Eventually this workaround PR shoud be redone and a proper runtime build should be setup that will then also compile Fortran MODULE files.

Fixes #89403

mjklemm committed:
Compile and link ISO_FORTRAN_ENV

mjklemm committed:
Fixes #89403

Moves definitions of the kind arrays into a Fortran MODULE to not only
emit the MOD file, but also compile that MODULE file into an object
file.  This file is then linked into libFortranRuntime.so.

Eventually this workaround PR shoud be redone and a proper runtime build
should be setup that will then also compile Fortran MODULE files.

klausler commented on flang/module/__fortran_builtin_kinds.f90:
```diff
@@ -0,0 +1,110 @@
+!===-- module/__fortran_bultin_kinds.f90 --=--------------------------------===!
```
why can't these definitions remain in iso_fortran_env?

mjklemm replied:
Flang was giving me compiler errors for numeric_storage_size, which seems it needs to not be compiled as part of the module file but rather as part of when the module has been included into the host routine.

That's I basically went back to my original idea of how I wanted to fix the bug and moved the intrinsic definitions into their own module file.

klausler replied:
__builtin_numeric_storage_size() gets special treatment in folding that can easily be adjusted without this kind of disruption.

mjklemm replied:
Can you provide a hint how this could be done?  I'm not that deep into the inner workings for Flang yet to see how this can be done.

klausler replied:
Grep for `builtin_numeric_storage` and you'll see what's going on.

mjklemm replied:
I have done that already.  Thing is: when the context is right, `__build_numeric_storage_size` is immediately folded.  However, then compiling `iso_fortran_env.f90`, line 1324-1326 in `fold-integer.cpp` simply leave the evaluation of `__builtin_numeric_storage_size` in the intermediate code, so that the code-gen picks it up and utterly complains about not having code-gen for it.  So, I have tried to add that, but that does not seem the right path to me.

mjklemm replied:
Ping :-)

klausler replied:
yes?

mjklemm replied:
Is my sketched approach the correct one or is there a simpler and better way to keep the original expression when compiling the module file?

klausler replied:
I'm confident that the problem can be solved without splitting the module source file by properly defining the compilations in which the folding the intrinsic must be deferred.

mjklemm replied:
Honestly, I did not see how.  So, hints or even a patch about how to do this will be appreciated.

klausler replied:
In `flang/lib/Evaluate/fold-integer.cpp` ca. line 1340 is the code that folds (underscore underscore) `builtin_numeric_storage`.  It defers that folding today until the reference to that intrinsic function is coming from a module file, at which point the size of a default numeric storage unit must be known.  If you're changing things so that `ISO_FORTRAN_ENV` is now going to be compiled with the size of a default numeric storage unit being well-defined, then you can probably just remove the code that defers the folding.

mjklemm replied:
When I remove the code, the compiler folds the current value `32_4` into the `iso_fortran_env.mod` file.  From there on, `numeric_storage_size` always reports 32 as the storage size, regardless of compiler options.

The code removed:

```C++
    if (!context.moduleFileName()) {
      // Don't fold this reference until it appears in the module file
      // for ISO_FORTRAN_ENV -- the value depends on the compiler options
      // that might be in force.
    } else {
```

klausler replied:
Yes, but since you now need to also compile the modules into distinct `*.o` files for each combination of target and relevant compilation options, isn't that what has to happen?

mjklemm replied:
That's not needed, if we do the separation like I'm proposing in this PR.  Everything that is compiled into an object file is independent of the relevant compiler options (like `-fdefault-real-8`), while `numeric_storage_size` is left in `iso_fortran_env` that is only treated with `-fsyntax-only`, which resolves the whole issue w/o touching the compiler itself.

klausler replied:
And complicated module file shenanigans are better somehow than "touching" the compiler?  I guess that's a matter of opinion.  I'll let you find some other reviewer for this, but you can help them out by at least documenting what's going on here with some comments, and fixing the name of the file in its header.

clementval commented:
Did you move the definitions because you get an error compiling the full `iso_fortran_env.f90` file? I think we can just wrap the part used during folding in a macro and not include it when we build the object file.

clementval commented on flang/tools/flang-driver/CMakeLists.txt:
```diff
@@ -19,8 +19,6 @@ add_flang_tool(flang-new
   # These libraries are used in the linker invocation generated by the driver
   # (i.e. when constructing the linker job). Without them the driver would be
   # unable to generate executables.
```
You can probably also remove the comment

mjklemm replied:
Good point.  Will do.

mjklemm replied:
Done

sscalpone commented:
Thanks for looking at this, @mjklemm !   My thought was to move the build of the module files to be part of the runtime build.     One advantage is that the object file can be put into the libFortranRuntime.  The other is that the module file build is more closely associated with the target instead of the host.

mjklemm commented:
I consider this a first step to resolve the bug, albeit with a workaround.  In the long run, I totally agree that the Fortran runtime part should be built like any other LLVM runtime.

As part of that the module file should be considered to move to be within in the runtime folder and also the CMake for it should move from tools/f18 into the runtime folder.  I think that would greatly simplify the CMake code needed.

mjklemm commented:
> Did you move the definitions because you get an error compiling the full `iso_fortran_env.f90` file? I think we can just wrap the part used during folding in a macro and not include it when we build the object file. 
> 
> 
> 
> This is the part of the module I'm talking about:
> 
> 
> 
> ```
> 
>   intrinsic :: __builtin_numeric_storage_size
> 
>   ! This value depends on any -fdefault-integer-N and -fdefault-real-N
> 
>   ! compiler options that are active when the module file is read.
> 
>   integer, parameter, public :: numeric_storage_size = &
> 
>     __builtin_numeric_storage_size()
> 
> ```

Yes, exactly.  I think we can leave this as is if we feel that moving the compiled part of iso_fortran_env into it's own module file like I did.

Not sure if we would need this for other stuff in the future, but maybe splitting the intrinsic modules into an implementation file and a definitions file might be something to consider going forward.

mjklemm committed:
Remove obsolete comment

mjklemm committed:
Set the CACHE variable only once

Meinersbur commented on flang/tools/f18/CMakeLists.txt:
```diff
@@ -58,18 +76,41 @@ if (NOT CMAKE_CROSSCOMPILING)
+  # Set a CACHE variable that is visible to the CMakeLists.txt in runtime/, so that
+  # the compiled Fortran modules can be added to the link line of the FortranRuntime
+  # library.
+  set(FORTRAN_MODULE_OBJECTS ${module_objects} CACHE INTERNAL "")
```
```suggestion
  set(FORTRAN_MODULE_OBJECTS ${module_objects} CACHE INTERNAL "" FORCE)
```
Without the [`FORCE`](https://cmake.org/cmake/help/latest/command/set.html#set-cache-entry) flag, the cached value will only be overwritten if it was not previously listed in `CMakeCache.txt`. So if a patch ever adds another module_object, all previous CMake build directories must be deleted.

A better alternative is [`set_property(GLOBAL PROPERTY ...)`](https://cmake.org/cmake/help/latest/command/set_property.html). `set(... CACHE)` does not use any caching and works as long as `set_property` and `get_property` are evaluated in the expected order (this file before `runtime/CMakeLists.txt`).

clementval commented on flang/module/iso_fortran_env_impl.f90:
```diff
@@ -0,0 +1,114 @@
+!===-- module/__fortran_bultin_kinds.f90 --=--------------------------------===!
```
```suggestion
!===-- module/iso_fortran_env_impl.f90 --=--------------------------------===!
```

mjklemm committed:
Update iso_fortran_env_impl.f90

Co-authored-by: Valentin Clement (バレンタイン クレメン) <clementval@gmail.com>

mjklemm committed:
Update flang/tools/f18/CMakeLists.txt

Co-authored-by: Michael Kruse <github@meinersbur.de>

mjklemm commented:
I have made the requested updates.

mjklemm commented:
@kkwli @Meinersbur @clementval Ping :-)

Meinersbur commented on flang/tools/f18/CMakeLists.txt:
```diff
@@ -58,18 +76,41 @@ if (NOT CMAKE_CROSSCOMPILING)
+    # If a module has been compiled into an object file, add the file to
+    # the link line for the FortranRuntime library.
+    if(include_in_link)
+      set(module_objects ${module_} ${object_output})
```
```suggestion
      list(APPEND module_objects ${module_})
```

Meinersbur commented on flang/module/iso_fortran_env_impl.f90:
```diff
@@ -0,0 +1,114 @@
+      pack([real64x2], real64x2 >= 0), &
+      pack([real128], real128 >= 0) &
+    ]
+end module iso_fortran_env_impl
```
[nit] Missing newline at end of file (UNIX convention; C and C++ before C++11 even require it)

Meinersbur reviewed (approved):
LGTM, with two nits.

Meinersbur commented:
Btw, you should remove the draft status so flang category subscribers are notified

mjklemm committed:
Update flang/tools/f18/CMakeLists.txt

Co-authored-by: Michael Kruse <github@meinersbur.de>

mjklemm committed:
Add newline

mjklemm committed:
Fix typo that came in from suggestion

mjklemm committed:
Add test

The test compiles and links a short program that uses the three kind
arrays.  It is supposed to fail and link time if the symbols are not
provided in the runtime.

kkwli reviewed (approved):
LG

mjklemm committed:
Revert "Add test"

This reverts commit 04f52d65d4a3691c92a9cb9908535fdd64aa18ba.

vzakhari commented on flang/runtime/CMakeLists.txt:
```diff
@@ -169,6 +169,7 @@ set(sources
   unit-map.cpp
   unit.cpp
   utf.cpp
+  ${FORTRAN_MODULE_OBJECTS}
```
This breaks some of my builds.  I believe this won't work in general unless you make sure that the object file is built before `FortranRuntime` target tries to build, i.e. there has to be a target dependency in this CMake file.

I think you can try to reproduce it by deleting the module object file and running `make FortranRuntime` - it will fail with something like:
```
No rule to make target `tools/flang/tools/f18/iso_fortran_env_impl.o', needed by `lib/libFortranRuntime.a'.  Stop.
```

mjklemm replied:
On it.  Seems like Ninja understands the dependency, as it has a single build file that seems the generated object file in one place and defers the target in the runtime libraries until that object file has been built.  This does not seem to be the case for Makefile builds, because they are split across different Makefiles.

mjklemm replied:
Created https://github.com/llvm/llvm-project/pull/99737 to hopefully resolve this.  Can you please see if the PR resolves the issue?

And sorry for breaking your builds!

vzakhari replied:
Thank you, Michael.  No problem, it should fix my builds, and building `module_files` target before the complete build was a workaround for me.
